### DISCUSSION TITLE
To ASGIMiddleware or not

### DISCUSSION THREAD

**Discussion by dineshbvadhia** [encode/starlette#2730] [Q&A]
The asgiref sync-to-async and async-to-sync function decoraters are used extensively in our code. After reading the Starlette pure-asgi-middleware documentation https://www.starlette.io/middleware/#pure-asgi-middleware, should we be using the example asgiref-specific ASGIMIddleware?

**Comment by adriangb** [comment]
I don't get how those decorators are related to the choice of middleware.

But generally to minimize issues I would:
- Do as much in async code as you can
- Minimize and isolate sync code and run in it in worker threads. Be careful of thread limits, you can cause deadlocks.
- Be careful with blocking the event loop by running sync code in it.
- Use pure ASGI middleware instead of `BaseHTTPMiddelware` or the `@app.middleware` decorator. If you're using FastAPI async dependencies are a good replacement for one-off simple middleware.

**Reply by dineshbvadhia** [reply]
Using asgiref, my code already follows:

- Do as much in async code as you can
- Minimize and isolate sync code and run in it in worker threads. Be careful of thread limits, you can cause deadlocks.
- Be careful with blocking the event loop by running sync code in it.

The question is: Do I also need to configure Starlette with the example asgiref-specific ASGIMIddleware or is this a silly question?

**Reply by adriangb** [reply]
> Using asgiref, my code already follows:

Using asgiref or not doesn't guarantee you are following those principles.

> Do I also need to configure Starlette with the example asgiref-specific ASGIMIddleware or is this a silly question?

What is "example asgiref-specific ASGIMIddleware"?

**Reply by dineshbvadhia** [reply]
Going to close the discussion because I need to get my head around Starlette's ASGIMiddleware. Thanks for hanging in :)
